### PR TITLE
Feature/test box metadata

### LINKED ISSUE #12: push box art to ipfs
the art should exist for addresses populated in #11

### PR DESCRIPTION
**Pull request by AddressXception**
fixes #12 

makes incremental progress on the bonding curve

remove the AllowList hardcode and instead let be set by admin

this PR is used as a release candidate

### CODE REVIEW

File: contracts/token/RarePizzasBox.sol [hunk 1]
```diff
@@ -88,18 +100,19 @@ contract RarePizzasBox is
 
     function purchase() public payable virtual override {
         require(
-            block.timestamp >= publicSaleStart_timestampInS || allowed(msg.sender),
-            "RAREPIZZA: The sale hasn't started yet"
+            block.timestamp >= publicSaleStart_timestampInS ||
```
Comment (CryptoSpaces) [inline contracts/token/RarePizzasBox.sol:103]: Could we split this up in 2 requires?

File: contracts/token/RarePizzasBox.sol [hunk 2]
```diff
@@ -88,18 +100,19 @@ contract RarePizzasBox is
...
         );
-        require(totalSupply().add(1) <= MAX_TOKEN_SUPPLY, 'RAREPIZZA: purchase would exceed maxSupply');
+        require(totalSupply().add(1) <= MAX_TOKEN_SUPPLY, 'RAREPIZZA: exceeds supply.');
 
         uint256 price = getPrice();
-        require(price == msg.value, 'RAREPIZZA: price must be on the curve');
+        require(msg.value >= price, 'RAREPIZZA: price too low');
 
+        // presale addresses can purchase up to 10 total
+        _presalePurchaseCount[msg.sender] += 1;
```
Comment (CryptoSpaces) [inline contracts/token/RarePizzasBox.sol:113]: Should we make this countable to?

File: contracts/token/RarePizzasBox.sol [hunk 3]
```diff
@@ -114,68 +127,61 @@ contract RarePizzasBox is
...
-        return bitcoinPriceInWei;
-    }
+    // IRarePizzasBoxAdmin
 
-    /**
-     * allows the contract owner to mint up to a specific number of boxes
-     */
-    function mint(address to, uint8 count) public virtual onlyOwner {
-        require(count > 0, 'RAREPIZZA: must provide a count');
+    function mint(address toPizzaiolo, uint8 count) public virtual override onlyOwner {
```
Comment (CryptoSpaces) [inline contracts/token/RarePizzasBox.sol:136]: Can owner mint to any address? Should these not be assigned to multi sig wallet?

File: contracts/token/RarePizzasBox.sol [hunk 4]
```diff
@@ -114,68 +127,61 @@ contract RarePizzasBox is
...
-     */
-    function updateBitcoinPriceInWei(int256 fallbackValue) public virtual onlyOwner {
+    function setPresaleAllowed(uint8 count, address[] memory toPaisanos) public virtual override onlyOwner {
+        for (uint256 i = 0; i < toPaisanos.length; i++) {
+            require(toPaisanos[i] != address(0), 'RAREPIZZA: dont be silly');
+            _presaleAllowed[toPaisanos[i]] = count;
+        }
+    }
+
+    function setSaleStartTimestamp(uint256 epochSeconds) public virtual override onlyOwner {
```
Comment (CryptoSpaces) [inline contracts/token/RarePizzasBox.sol:170]: This could be used to halt sales and allow whitelist people to buy only

**Review by CryptoSpaces — COMMENTED** [review]
Great work 👍

File: contracts/token/RarePizzasBox.sol [hunk 5]
```diff
@@ -114,68 +127,61 @@ contract RarePizzasBox is
...
-        _safeMint(to, id);
-        _assignBoxArtwork(id);
+        _internalMintWithArtwork(toPaisano);
     }
 
-    /**
-     * allows the owner to update the cached bitcoin price
-     */
-    function updateBitcoinPriceInWei(int256 fallbackValue) public virtual onlyOwner {
+    function setPresaleAllowed(uint8 count, address[] memory toPaisanos) public virtual override onlyOwner {
```
Comment (CryptoSpaces) [inline contracts/token/RarePizzasBox.sol:163]: Could we make this function and the setSaleStart to be callable only once or twice?

File: contracts/token/RarePizzasBox.sol (hunk 3 above)
Comment (AddressXception) [reply contracts/token/RarePizzasBox.sol:144]: owner can mint to any address yes

File: contracts/token/RarePizzasBox.sol (hunk 4 above)
Comment (AddressXception) [reply contracts/token/RarePizzasBox.sol:180]: yes correct that is the desired behavior

File: contracts/token/RarePizzasBox.sol (hunk 5 above)
Comment (AddressXception) [reply contracts/token/RarePizzasBox.sol:173]: dont want ot limit it arbitrarily in that way

File: contracts/token/RarePizzasBox.sol [hunk 6]
```diff
@@ -110,13 +110,15 @@ contract RarePizzasBox is
...
         payable(msg.sender).transfer(msg.value - price);
 
-        // presale addresses can purchase up to 10 total
+        // presale addresses can purchase up to X total
         _presalePurchaseCount[msg.sender] += 1;
         _purchased_pizza_count.increment();
         _internalMintWithArtwork(msg.sender);
 
         // BUY ONE GET ONE FREE!
         if (totalSupply().add(1) == MAX_TOKEN_SUPPLY) {
```
Comment (CryptoSpaces) [inline contracts/token/RarePizzasBox.sol:119]: Could we use `_purchased_pizza_count.add(1) == MAX_PURCHASABLE_SUPPLY`

This only relates to the purchasable boxes and should probably be separated from the mintable boxes

File: contracts/token/RarePizzasBox.sol (hunk 1 above)
Comment (AddressXception) [reply contracts/token/RarePizzasBox.sol:103]: not necessary to incur the second check if the sale is already on

File: contracts/token/RarePizzasBox.sol (hunk 6 above)
Comment (AddressXception) [reply contracts/token/RarePizzasBox.sol]: updated

File: contracts/token/RarePizzasBox.sol (hunk 2 above)
Comment (AddressXception) [reply contracts/token/RarePizzasBox.sol:114]: skipping based on convo

**Review by CryptoSpaces — APPROVED** [review]

**Pull request merged**